// parce/.eslintrc.js
/* eslint-disable */

module.exports = {
    extends: ['eslint:recommended', 'plugin:vue/base', 'airbnb-base', 'plugin:prettier/recommended', 'prettier/standard'],
    parserOptions: {
        parser: 'babel-eslint',
        ecmaVersion: 6
    },
    plugins: ['prettier', 'html', 'import'],
    env: {
        browser: true,
        node: true
    },
    globals: {},
    rules: {
        'prettier/prettier': 'error',
        indent: [2, 4, { SwitchCase: 1 }], // 缩进
        'no-console': [2],
        'max-len': [0], // 一行最大的代码量
        'comma-dangle': [2, 'never'], // 尾逗号
        'arrow-parens': [2, 'as-needed'], // 箭头函数参数 括号
        'no-param-reassign': [2, { props: false }], // 不要改写函数参数
        'arrow-body-style': [0], // 箭头函数的返回值与大括号 () => {...} vs () => ...
        'object-curly-newline': [0], // 对象换行
        'global-require': [0],
        'prefer-const': [
            2,
            {
                destructuring: 'all'
            }
        ],
        'object-curly-spacing': [2, 'always'],
        'no-underscore-dangle': [0], // 变量下划线
        'func-names': [2, 'never'], // 函数名
        'import/no-unresolved': [0],
        'import/no-dynamic-require': [0],
        'import/no-extraneous-dependencies': [
            2,
            {
                devDependencies: true,
                peerDependencies: true
            }
        ]
    }
};
